File: 11/day11.py
# https://adventofcode.com/2024/day/11
from collections import Counter

def main(file_path: str) -> None:
    counter = Counter()
    with open(file_path) as f:
        for line in f:
            for num in [int(x) for x in line.split()]:
                counter[num] += 1
                
    def count_stones(num_blinks: int) -> None:
        curr_counter = counter.copy()
        for blink in range(num_blinks):
            new_counter = Counter()
            for k, v in curr_counter.items():
                if k == 0:
                    new_counter[1] += v
                else:
                    s = str(k)
                    if len(s) % 2 == 0:
                        k1 = int(s[:len(s)//2])
                        k2 = int(s[len(s)//2:])
                        new_counter[k1] += v
                        new_counter[k2] += v
                    else:
                        new_counter[k*2024] += v
                    
                # the old stone has been replaced
                curr_counter[k] -= v
            curr_counter.update(new_counter)
            # print_blink(blink, dict(counter))
        print(sum(curr_counter.values()))
    
    count_stones(25)
    count_stones(75)
    
def print_blink(b, d):
    print(f'{b+1}: ', end=' ')
    for k, v in d.items():
        if v > 0:
            print(f'{k}: {v}', end=',')
    print()
if __name__ == '__main__':
    # main('./11/input0.txt')
    main('./11/input.txt')

File: 11/day11.js
const fs = require('fs');

function main(filePath) {
  const data = fs.readFileSync(filePath, "utf8").split("\n");
  const counter = {};
  for (v of data[0].split(' ')) { // the last line is an empty line 
    counter[v] = (counter[v] || 0) + 1;
  }

  function countStones(counter, blinks) {
    const clone = {...counter};
    for (let _ = 0; _ < blinks; _++) {
      for (const [key, v] of Object.entries(clone)) {
        const k = Number(key);  // since key is of type string, convert to number
        if (k === 0) {
          clone[1] = (clone[1] || 0) + v;
        } else {
          if (key.length % 2 === 0) {
            const mid = Math.floor(key.length / 2);
            const h1 = Number(key.substring(0, mid));
            const h2 = Number(key.substring(mid, key.length));
            clone[h1] = (clone[h1] || 0) + v;
            clone[h2] = (clone[h2] || 0) + v;
          } else {
            clone[k * 2024] = (clone[k * 2024] || 0) + v;
          }
        }
        clone[key] -= v;
      }
    }
    const ans = Object.values(clone).reduce((acc, v) => acc + v, 0);
    console.log(ans);
  }

  countStones(counter, 25);
  countStones(counter, 75);
}

// main('./11/input0.txt')
main('./11/input.txt')
